### DISCUSSION TITLE
Is getStaticProps ran at server side or build time?

### DISCUSSION THREAD

**Discussion by sontek** [vercel/next.js#26054] [Help]
I'm currently getting familiar with next.js through this tutorial:

https://nextjs.org/learn/basics/data-fetching/getstaticprops-details

I thought I understood what it was saying until two statements that felt like they opposed each other:

1. It says that `getStaticProps` is run server side and so it will have access to things like the database:

 
> This is possible because getStaticProps only runs on the server-side. It will never run on the client-side. It won’t even be included in the JS bundle for the browser. That means you can write code such as direct database queries without them being sent to browsers.

2. It follows it up with the next sentence saying:

> In production, getStaticProps runs at build time. However, this behavior can be enhanced using the fallback key returned by getStaticPaths

So I was hoping to gain some clarity.   My build server (Github Actions) isn't going to have access to the database, so doing queries won't be available at `build time` but they will be available `server-side`.    I understand there is `getServerSideProps` as well but those are unique *per request*.    I was hoping to use `getStaticProps` for initial data that is the same for all requests but it needs to be fetched on the server not at build time.

**Comment by akellbl4** [comment]
Hi @sontek, I'll try to clarify how usually these methods work.

First of all `getStaticProps` was created for Static Site Generation. It means all of the pages of a site should be created at the build stage and data for the pages is fetched at this stage. As a result, we have a fully static site with HTML/CSS/JS.

Then we got Incremental Static Generation. It enables you to rebuild the pages that were already deployed and create new ones if they appeared in your database. This feature configures with two params first is `revalidate` and refers to the time that tells when a page should be re-rendered. The second param is `fallback`. It tells when re-render or to create a new page (`true` in background, `blocking` as SSR).

I assume you need to build pages for some of the same type of data (for example product pages or blog posts).
Then you can use do something like this:

```js
export function getStaticPaths() {
  return {
    paths: [],
    fallback: 'blocking' // or `true` if you want to render a page on the first request to it on the client
  }
}

export async function getStaticProps() {
  const data = await fetchPageData()

  return {
    props: { data },
   revalidate: 60, // revalidate page every 60 seconds if request comes
  }
}
```

This example gives you the ability not to render pages on the build stage and deploy your app. Then after deploy pages will be rendered by request and will live for 60 seconds and re-renders by request.

I hope my explanation was clear enough.
